x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -805,65 +835,28 @@ private void innerRefresh(String tokenDocId, Map<String, Object> source, long se
-                        logger.info("could not find superseding token document [{}] for token document [{}], retrying",
-                                refreshTokenStatus.getSupersedingDocId(), tokenDocId);
-                        maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
-                    }
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    if (isShardNotAvailableException(e)) {
-                        logger.info("could not find superseding token document [{}] for refresh, retrying",
-                                refreshTokenStatus.getSupersedingDocId());
-                        maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
-                    } else {
-                        logger.warn("could not find superseding token document [{}] for refresh", refreshTokenStatus.getSupersedingDocId());
-                        onFailure.accept(invalidGrantException("could not refresh the requested token"));
-                    }
-                }
-            });
+                    tokenDocId, refreshTokenStatus.getSupersededBy());
+            getSupersedingTokenDocAsyncWithRetry(refreshTokenStatus, backoff, listener);
         } else {
             final String newUserTokenId = UUIDs.randomBase64UUID();
+            final Version newTokenVersion = getTokenVersionCompatibility();
             final Map<String, Object> updateMap = new HashMap<>();
             updateMap.put("refreshed", true);
             updateMap.put("refresh_time", clock.instant().toEpochMilli());
-            updateMap.put("superseded_by", getTokenDocumentId(newUserTokenId));
+            if (newTokenVersion.onOrAfter(VERSION_TOKENS_INDEX_INTRODUCED)) {
+                // the superseding token document reference is formated as: "<alias>|<document_id>" ; the alias points to a single index
+                // containing the document with the said id
+                updateMap.put("superseded_by",
+                        getTokensIndexManagerForVersion(newTokenVersion).aliasName() + "|" + getTokenDocumentId(newUserTokenId));
+            } else {
+                // preservers the format of the reference so that old nodes in a mixed cluster can still understand it
+                updateMap.put("superseded_by", getTokenDocumentId(newUserTokenId));
+            }
             assert seqNo != SequenceNumbers.UNASSIGNED_SEQ_NO : "expected an assigned sequence number";
             assert primaryTerm != SequenceNumbers.UNASSIGNED_PRIMARY_TERM : "expected an assigned primary term";
-            final UpdateRequestBuilder updateRequest = client.prepareUpdate(SECURITY_INDEX_NAME, SINGLE_MAPPING_NAME, tokenDocId)
+            final SecurityIndexManager refreshedTokenIndex = getTokensIndexManagerForVersion(refreshTokenStatus.getVersion());
```

Comment:
> I guess you named it this way because we determine which index to use based on the already refreshed token

Yes, it is the name of the index that contains the token that was refreshed.

> Can we decide on two names for the new and old version indices and use them throughout the TokenService?

This is a local variable and the name of the variable should reflect the meaning inside the function. For example, here, it is not the _old_ or the _new_ index, it is the _index that contains the token we have received and have to refresh_ , it could be either the _old_ or the _new_ index.

Reply:
Gotcha, you're right.